### PR TITLE
scripts: align sram .text section when userspace and code reloation enabled

### LINKED ISSUE #17038: code relocation generating different memory layout cause user mode not working
**Describe the bug**
In the magic building process for user mode, there are different memory layout as below log show. 
The map file of building zephyr_prebuilt.elf is:
```
 .text.sedi_pm_pmu2nvic_isr
                0x0000000060011188        0xa libbsp_sedi_arm.a(pm.c.obj)
                0x0000000060011188                sedi_pm_pmu2nvic_isr
 *(SORT_BY_ALIGNMENT(.gnu.linkonce.t.*))
 *(SORT_BY_ALIGNMENT(.glue_7t))
 .glue_7t       0x0000000060011192        0x0 linker stubs
 *(SORT_BY_ALIGNMENT(.glue_7))
 .glue_7        0x0000000060011192        0x0 linker stubs
 *(SORT_BY_ALIGNMENT(.vfp11_veneer))
 .vfp11_veneer  0x0000000060011192        0x0 linker stubs
 *(SORT_BY_ALIGNMENT(.v4_bx))
 .v4_bx         0x0000000060011192        0x0 linker stubs
                0x0000000060011192                _priv_stacks_text_area_start = .
 *(SORT_BY_ALIGNMENT(.priv_stacks.text*))
 *fill*         0x0000000060011192        0x2 
```

And ztest_thread_stack located in 0x60030000 from below log and we can also find it from map file.
`gen_kobject_list.py: symbol 'ztest_thread_stack' at 0x60030000 contains 1 object(s)`

The map file of building priv_stacks_prebuilt.elf is:
```
.text.sedi_pm_pmu2nvic_isr
                0x0000000060011188        0xa libbsp_sedi_arm.a(pm.c.obj)
                0x0000000060011188                sedi_pm_pmu2nvic_isr
 *fill*         0x0000000060011192        0x6 
 .text.sedi_pm_pmu2nvic_isr.__stub
                0x0000000060011198       0x68 linker stubs
 *(SORT_BY_ALIGNMENT(.gnu.linkonce.t.*))
 *(SORT_BY_ALIGNMENT(.glue_7t))
 .glue_7t       0x0000000060011200        0x0 linker stubs
 *(SORT_BY_ALIGNMENT(.glue_7))
 .glue_7        0x0000000060011200        0x0 linker stubs
 *(SORT_BY_ALIGNMENT(.vfp11_veneer))
 .vfp11_veneer  0x0000000060011200        0x0 linker stubs
 *(SORT_BY_ALIGNMENT(.v4_bx))
 .v4_bx         0x0000000060011200        0x0 linker stubs
                0x0000000060011200                _priv_stacks_text_area_start = .
```
And ztest_thread_stack located in 0x60050000 from below log and we can also find it from map file.
`gen_priv_stacks.py: symbol 'ztest_thread_stack' at 0x60050000 contains 1 object(s)`

So that difference will cause the wrong hash table in priv_stacks_hash.c because of the different  .rodata .noinit .data sections in the same one sram region and it will cause the user mode doesn't work.

### PR DESCRIPTION
**Pull request by wentongwu**
1) Code relocation causes comipler randomly generates _stub that will
    cause different memory layout in the magic building process for user
    mode. And the different memory layout will generate wrong hash table
    in priv_stacks_hash.c and it will cause user mode doesn't work.
    Align sram .text section when userspace and code relocation enabled
    to generate same .rodata, .data and .bss sections in sram region.
    User can adjust Kconfig RELOCATION_SRAM_TEXT_ALIGN_SIZE to make less
    wasting memory.

    2) make mpu align inside sections instead of outside to
    avoid overlap.

    3) fix spell typo in target_relocation.cmake.

Fixes: #17038.

### CODE REVIEW

**Comment by wentongwu** [comment]
@andrewboie @ioannisg @nashif could you please take a look this change?Thanks

**Comment by andrewboie** [comment]
I think I understand this patch, can you describe in a bit more detail how the memory map is shifting before this fix was applied?

**Comment by wentongwu** [comment]
> I think I understand this patch, can you describe in a bit more detail how the memory map is shifting before this fix was applied?

@andrewboie please see issue #17038 .

**Comment by wentongwu** [comment]
push again to solve conflict with master.

**Comment by andrewboie** [comment]
OK, thanks for the further details.
If I understand correctly, the issue is that the ".text.sedi_pm_pmu2nvic_isr.__stub" gets added to the text section in zephyr.elf, but not zephyr_prebuilt.elf. Something is causing this to be pulled into only the final binary. At this time it is not known what is causing this to happen. This patch is a workaround to reserve some room in the prebuilt binary for text area expansion, so that the memory addresses don't change in between the two builds.

**Comment by wentongwu** [comment]
shippable fail not related to my change
1. up_squared/samples/boards/up_squared/gpio_counter/sample.board.up_squared.gpio_counter   failure: build_error
2. qemu_xtensa/tests/kernel/sched/preempt/kernel.sched.preempt   failure: timeout
3. up_squared/samples/synchronization/sample.synchronization failure: build_error

Re-submit.

**Comment by wentongwu** [comment]
> OK, thanks for the further details.
> If I understand correctly, the issue is that the ".text.sedi_pm_pmu2nvic_isr.__stub" gets added to the text section in zephyr.elf, but not zephyr_prebuilt.elf. Something is causing this to be pulled into only the final binary. At this time it is not known what is causing this to happen. This patch is a workaround to reserve some room in the prebuilt binary for text area expansion, so that the memory addresses don't change in between the two builds.

yes, that's workaround for it so far. 
could you please take a look this change again, there is no code change for the re-submit, just to pass the shippable, and someone is waiting this change. Thanks a lot. @andrewboie

**Comment by andrewboie** [comment]
What steps have you taken to isolate why the extra data in the text section is being generated for zephyr.elf and not zephyr_prebuilt.elf? Where are these symbols coming from? Are they somehow being pulled in due to the gperf code we add in this phase?

> someone is waiting for this change

If someone is waiting on this patch urgently they can cherry-pick it.

File: Kconfig.zephyr [hunk 1]
```diff
@@ -82,6 +82,15 @@ config CODE_DATA_RELOCATION
 	  files should be specified in the CMakeList.txt file with
 	  a cmake API zephyr_code_relocation().
 
+config RELOCATION_SRAM_TEXT_ALIGN_SIZE
+	int "code relocation sram text section align size"
+	default 80000
+	depends on ARCH_HAS_USERSPACE && CODE_DATA_RELOCATION
+	help
+	  Align sram text section when userspace and code relocation
```
Comment (andrewboie) [inline Kconfig.zephyr]: This help text really doesn't explain what is going on. You need to expand on this.

File: Kconfig.zephyr [hunk 2]
```diff
@@ -82,6 +82,15 @@ config CODE_DATA_RELOCATION
...
+	  enabled to generate same .rodata, .data and .bss sections
+	  in sram region for the magic building process.
```
Comment (andrewboie) [inline Kconfig.zephyr]: "magic building process" is NOT a technical term

File: Kconfig.zephyr [hunk 3]
```diff
@@ -82,6 +82,15 @@ config CODE_DATA_RELOCATION
 	  files should be specified in the CMakeList.txt file with
 	  a cmake API zephyr_code_relocation().
 
+config RELOCATION_SRAM_TEXT_ALIGN_SIZE
+	int "code relocation sram text section align size"
+	default 80000
```
Comment (andrewboie) [inline Kconfig.zephyr]: What does this mean? We are by default reserving 80 Kilobytes of space? Why so much, given that the offset in the binary was only 8 bytes? The help text needs to clearly explain the semantics on the value we set here.

File: cmake/linker/ld/target_relocation.cmake [hunk 4]
```diff
@@ -3,28 +3,31 @@
 # See root CMakeLists.txt for description and expectations of these macros
 
 macro(toolchain_ld_relocation)
-  set(MEM_RELOCATAION_LD   "${PROJECT_BINARY_DIR}/include/generated/linker_relocate.ld")
-  set(MEM_RELOCATAION_SRAM_DATA_LD
+  set(MEM_RELOCATION_LD   "${PROJECT_BINARY_DIR}/include/generated/linker_relocate.ld")
```
Comment (andrewboie) [inline cmake/linker/ld/target_relocation.cmake]: Did you make some whitespace changes or something? I can't tell the difference between the before and after, it looks like the code is unchanged.

File: cmake/linker/ld/target_relocation.cmake [hunk 5]
```diff
@@ -3,28 +3,31 @@
...
+    -o ${MEM_RELOCATION_LD}
+    -s ${MEM_RELOCATION_SRAM_DATA_LD}
+    -b ${MEM_RELOCATION_SRAM_BSS_LD}
+    -a ${MEM_RELOCATION_SRAM_TEXT_ALIGN_LD}
+    -c ${MEM_RELOCATION_CODE}
     DEPENDS app kernel ${ZEPHYR_LIBS_PROPERTY}
     )
 
-  add_library(code_relocation_source_lib  STATIC ${MEM_RELOCATAION_CODE})
+  add_library(code_relocation_source_lib  STATIC ${MEM_RELOCATION_CODE})
```
Comment (andrewboie) [inline cmake/linker/ld/target_relocation.cmake]: same as above, why is there even a diff?

File: cmake/linker/ld/target_relocation.cmake (hunk 4 above)
Comment (andrewboie) [reply cmake/linker/ld/target_relocation.cmake]: Disregard, I see it now

**Comment by wentongwu** [comment]
> What steps have you taken to isolate why the extra data in the text section is being generated for zephyr.elf and not zephyr_prebuilt.elf? Where are these symbols coming from? Are they somehow being pulled in due to the gperf code we add in this phase?

@andrewboie @jocelyn-li  @nashif One method I used is to put *stub to one separate section and do small align, but it doesn't work. Another idea is trying to put the .text section behind .data section, so the extra code will not impact .data section, I didn't do this one in afraid of broken other boards. And finally it's the workaround method, it works well, but now I know it doesn't fit the opensource spirit, I'm sorry for that. 

Look at it again today, I find the extra stubs are comprised of veneer functions, that's because the distance between calling code and the called function is too far(because of relocation, also there are other reasons to cause the far distance), the linker will automatically generate and insert veneer functions. Now I'm trying to figure out the solution.  And now I want to close this PR.

**Pull request closed without merging**